# Top Spotify songs 

In [2]:
import numpy as np
import pandas as pd

spotify = pd.read_csv("spotify.csv")

In [10]:
spotify.head()

id                          name       artists  \
0  6DCZcSspjsKoFjzjrWoCd                    God's Plan         Drake   
1  3ee8Jmje8o58CHK66QrVC                          SAD!  XXXTENTACION   
2  0e7ipj03S05BNilyu5bRz    rockstar (feat. 21 Savage)   Post Malone   
3  3swc6WTsr7rl9DqQKQA55  Psycho (feat. Ty Dolla $ign)   Post Malone   
4  2G7V7zsVDxg1yRsu7Ew9R                In My Feelings         Drake   

   danceability  energy  key  loudness  mode  speechiness  acousticness  \
0         0.754   0.449  7.0    -9.211   1.0       0.1090        0.0332   
1         0.740   0.613  8.0    -4.880   1.0       0.1450        0.2580   
2         0.587   0.535  5.0    -6.090   0.0       0.0898        0.1170   
3         0.739   0.559  8.0    -8.011   1.0       0.1170        0.5800   
4         0.835   0.626  1.0    -5.833   1.0       0.1250        0.0589   

   instrumentalness  liveness  valence    tempo  duration_ms  time_signature  
0          0.000083     0.552    0.357   77.169     198973.0             4.0  
1          0.003720     0.123    0.473   75.023     166606.0             4.0  
2          0.000066     0.131    0.140  159.847     218147.0             4.0  
3          0.000000     0.112    0.439  140.124     221440.0             4.0  
4          0.000060     0.396    0.350   91.030     217925.0             4.0

In [9]:
spotify.describe()

danceability      energy         key    loudness        mode  \
count     100.00000  100.000000  100.000000  100.000000  100.000000   
mean        0.71646    0.659060    5.330000   -5.677640    0.590000   
std         0.13107    0.145067    3.676447    1.777577    0.494311   
min         0.25800    0.296000    0.000000  -10.109000    0.000000   
25%         0.63550    0.562000    1.750000   -6.650500    0.000000   
50%         0.73300    0.678000    5.000000   -5.566500    1.000000   
75%         0.79825    0.772250    8.250000   -4.363750    1.000000   
max         0.96400    0.909000   11.000000   -2.384000    1.000000   

       speechiness  acousticness  instrumentalness    liveness     valence  \
count   100.000000    100.000000        100.000000  100.000000  100.000000   
mean      0.115569      0.195701          0.001584    0.158302    0.484443   
std       0.104527      0.220946          0.013449    0.111662    0.206145   
min       0.023200      0.000282          0.000000    0.021500    0.079600   
25%       0.045350      0.040225          0.000000    0.094675    0.341000   
50%       0.074950      0.109000          0.000000    0.118500    0.470500   
75%       0.137000      0.247750          0.000031    0.170750    0.641500   
max       0.530000      0.934000          0.134000    0.636000    0.931000   

            tempo    duration_ms  time_signature  
count  100.000000     100.000000          100.00  
mean   119.904180  205206.780000            3.98  
std     28.795984   40007.893404            0.20  
min     64.934000   95467.000000            3.00  
25%     95.730750  184680.000000            4.00  
50%    120.116000  205047.500000            4.00  
75%    140.022750  221493.250000            4.00  
max    198.075000  417920.000000            5.00

## Number of artists in dataset

In [16]:
artist_count = spotify["artists"].nunique()
artist_count

70

There are 70 artists represented in the data set. 

## Mean danceability 

In [19]:
danceability_mean = spotify["danceability"].mean()
danceability_mean

0.71646

The mean danceability of the data set is 0.71646. 

## Defining feature set for the most "fun" song

I personally dislike heavy metal music, even if it is "fun", but some people like it. 

As "fun" is subjective and can be in every genre of music, the concept of genres should be dissacosiated with how fun the song is. This is so that the result can be as general as possible. Therefore, my philosophy of choosing the characteristics of a fun song is not based on my pre-conceived notion of what"fun" songs should be like. Rather, it is based on scientific evidence that some music characteristics evokes the necessary human emotions for a song to be considered "fun". 

With that in mind, I chose the following characteristics. Attached is a quick justification, after doing some research. 

Energy: High energy songs tend to be uplifting and a confidence booster. 

Mode: Songs played in a major key are judged as happy, while songs in minor key are judged as sad. (2)

Valence: Based on the data description given: "Tracks with high valence sound more positive (e.g. happy, cheerful, euphoric), while tracks with low valence sound more negative (e.g. sad, depressed, angry)". 

Tempo: "Stressing, Expressive and Amusing are the words gotten from a high tempo, whilst the opposite terms Relaxing, Expressionless and Boring are obtained for a low tempo." (1)

Other characteristics seem to be associated with genres (danceability, accousticness, loudness, speechiness). Thus, their omission is necessary so that the result is not biased towards a specific genre. 


Sources: 
(1) https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4971092/
(2) https://www.nme.com/blogs/nme-blogs/the-science-of-music-why-do-songs-in-a-minor-key-sound-sad-760215

In [73]:
spotify_new = spotify[spotify.columns[[0,1,2,4,7,12,13]]]
spotify_new.head()

id                          name       artists  energy  \
0  6DCZcSspjsKoFjzjrWoCd                    God's Plan         Drake   0.449   
1  3ee8Jmje8o58CHK66QrVC                          SAD!  XXXTENTACION   0.613   
2  0e7ipj03S05BNilyu5bRz    rockstar (feat. 21 Savage)   Post Malone   0.535   
3  3swc6WTsr7rl9DqQKQA55  Psycho (feat. Ty Dolla $ign)   Post Malone   0.559   
4  2G7V7zsVDxg1yRsu7Ew9R                In My Feelings         Drake   0.626   

   mode  valence    tempo  
0   1.0    0.357   77.169  
1   1.0    0.473   75.023  
2   0.0    0.140  159.847  
3   1.0    0.439  140.124  
4   1.0    0.350   91.030

In [74]:
spotify_new.describe()

energy        mode     valence       tempo
count  100.000000  100.000000  100.000000  100.000000
mean     0.659060    0.590000    0.484443  119.904180
std      0.145067    0.494311    0.206145   28.795984
min      0.296000    0.000000    0.079600   64.934000
25%      0.562000    0.000000    0.341000   95.730750
50%      0.678000    1.000000    0.470500  120.116000
75%      0.772250    1.000000    0.641500  140.022750
max      0.909000    1.000000    0.931000  198.075000

In [75]:
# To get rid of all songs played in minor key. 
spotify_new = spotify_new[spotify_new["mode"] != 0.0]

In [76]:
# To give equal weightage to the rest of the characteristics, each characteristics is divided into 5 bins. 
# Energy and valence have values between 0 to 1. Thus, it is easy to divide it into bins. 
spotify_new.loc[:,"energy"] = pd.cut(spotify_new["energy"], bins=[0, 0.2, 0.4, 0.6, 0.8,1], include_lowest=True, labels=[1,2,3,4,5]).values.astype('int64')
spotify_new.loc[:,"valence"] = pd.cut(spotify_new["valence"], bins=[0, 0.2, 0.4, 0.6, 0.8,1], include_lowest=True, labels=[1,2,3,4,5]).values.astype('int64')
# Tempo has values between 64 to 199. To divide into 5 bins, each bin would have an interval of 27. 
spotify_new.loc[:,"tempo"] = pd.cut(spotify_new["tempo"], bins=[64,91,118,145,172,199], include_lowest=True, labels=[1,2,3,4,5]).values.astype('int64')

In [77]:
spotify_new.head()

id                          name       artists  energy  \
0  6DCZcSspjsKoFjzjrWoCd                    God's Plan         Drake       3   
1  3ee8Jmje8o58CHK66QrVC                          SAD!  XXXTENTACION       4   
3  3swc6WTsr7rl9DqQKQA55  Psycho (feat. Ty Dolla $ign)   Post Malone       3   
4  2G7V7zsVDxg1yRsu7Ew9R                In My Feelings         Drake       4   
5  7dt6x5M1jzdTEt8oCbisT                    Better Now   Post Malone       3   

   mode  valence  tempo  
0   1.0        2      1  
1   1.0        3      1  
3   1.0        3      3  
4   1.0        2      2  
5   1.0        2      4

In [79]:
# Score adds up the bin number of energy, valence and tempo for each row. The row with the highest score can be said
# to be the most fun song. 
spotify_new["score"] = spotify_new["energy"].add(spotify_new["valence"]).add(spotify_new["tempo"])
spotify_new.loc[spotify_new["score"].idxmax()]

id                           1j6xOGusnyXq3l6IryKF3
name       D?jala que vuelva (feat. Manuel Turizo)
artists                                    Piso 21
energy                                           4
mode                                             1
valence                                          5
tempo                                            4
score                                           13
Name: 66, dtype: object

## Result

According to the aobve procedure, Dejala que vuelva by Piso 21 appears to be the most fun song of 2018. It scored highly in terms of energy, valence and tempo. 

I have not heard of the song before this exercise, but I can say that I enjoyed the song alot (and had alot of fun). 

## Potential imporvements

The metric used to evaluate the songs can be imporved upon. Energy, valence, tempo might not be equally important in contributing to the fun-ness of the song. 